proposals/esm-integration/EXAMPLES.md

```diff
@@ -0,0 +1,226 @@
+
+#### Example
+
+```
+// main.wasm
+(module
+  (import "./a.js" "memoryExport" (memory 0))
+)
+
+// a.js
+export {memoryExport} from "./b.wasm";
+
+// b.wasm
+(module
+  (memory 1)
+  (export "memoryExport" (memory 0))
+)
+```
+
+### JS <-> wasm cycle (where JS is higher in the module graph)
+
+#### JS exports
+| export type | value (not a WebAssembly.Global)* | global | memory | table | function |
+|-|-------------------------------|--------|--------|-------|----------|
+| | 0 if a const import and not in TDZ, otherwise Error | Error  | Error  | Error | snapshot if it is a function declaration, otherwise Error |
+
+#### wasm exports
+| export type | global       | memory       | table        | function     |
+|-|--------------|--------------|--------------|--------------|
+| | accurate binding | accurate binding | accurate binding | accurate binding |
+
+1. JS module is parsed
+1. wasm module is parsed
+1. wasm module has a lexical environment created for its exports. All exports are initially in TDZ.
+1. JS module is instantiated. All imports (including functions) from the wasm module are memory locations holding undefined.
+1. wasm module is instantiated and evaluated. Snapshots of imports are taken. Export bindings are initialized.
+1. JS module is evaluated. 
+
+#### Example
+
+```
+// a.js
+import {memoryExport} from "./b.wasm";
+export function functionExport() {
+    // do something with memory and DOM
+}
+
+// b.wasm
+(module
+  (memory 1)
+  (export "memoryExport" (memory 0))
+  (import "./a.js" "functionExport" (func $functionExport (result i32)))
+)
+```
+
+### wasm <-> JS cycle (where wasm is higher in the module graph)
+
+#### wasm exports
+| export type | global       | memory       | table        | function     |
+|-|--------------|--------------|--------------|--------------|
+| | accurate binding | accurate binding | accurate binding | accurate binding |
+
+#### JS exports
+| export type | value (not a WebAssembly.Global)* | global | memory | table | function |
+|-|-------------------------------|--------|--------|-------|----------|
+| | Error                         | snapshot  | snapshot  | snapshot | snapshot |
+
+1. wasm module is parsed
+1. JS module is parsed
+1. JS module is instantiated. 
+1. wasm module has a lexical environment created for its exports. All exports are initially in TDZ.
+1. JS module is evaluated. wasm exports lead to a ReferenceError if used.
+1. wasm module is instantiated and evaluated; wasm-exported bindings are updated to their appropriate JS API-exposed values. 
+
+#### Examples
+
+```
+// a.wasm
+(module
+  (memory 1)
+  (export "memoryExport" (memory 0))
+  (import "./b.js" "functionExport" (func $functionExport (result i32)))
+)
+
+// b.js
+import {memoryExport} from "./a.wasm";
+export function functionExport() {
+    // do something with memory and DOM
+}
+```
```

Comment:
☝️ Can you beef this example up. It's almost there but as is won't trigger an error.

Reply:
Why should it trigger an error?

Reply:
Was hoping to see 
> JS module is evaluated. wasm exports lead to a ReferenceError if used.

demonstrated.

Reply:
oh, just refer to memoryExport from outside of the function, to be run when the module executes.

But if I do that, there's nothing I can usefully export from this example! So it would have to be a separate example.